client/finality-grandpa-warp-sync/src/proof.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use codec::{Decode, Encode};
+
+use sc_finality_grandpa::{AuthoritySetChanges, GrandpaJustification};
+use sp_blockchain::Backend as BlockchainBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, NumberFor},
+};
+
+use crate::HandleRequestError;
+
+/// The maximum number of authority set change proofs to include in a single warp sync proof.
+const MAX_CHANGES_PER_WARP_SYNC_PROOF: usize = 256;
+
+#[derive(Decode, Encode)]
+pub struct AuthoritySetChangeProof<Block: BlockT> {
+	/// must contain the authorities for the following set
+	pub header: Block::Header,
+	/// this justification proves finality of the header above
+	/// it must be validated against the authorities of the previous set.
+	pub justification: GrandpaJustification<Block>,
+}
+
+#[derive(Decode, Encode)]
+pub struct WarpSyncProof<Block: BlockT> {
+	proofs: Vec<AuthoritySetChangeProof<Block>>,
+}
+
+impl<Block: BlockT> WarpSyncProof<Block> {
+	pub fn generate<Backend>(
+		backend: &Backend,
+		begin: Block::Hash,
+		set_changes: &AuthoritySetChanges<NumberFor<Block>>,
+	) -> Result<WarpSyncProof<Block>, HandleRequestError>
+	where
+		Backend: BlockchainBackend<Block>,
+	{
+		// TODO: cache best response (i.e. the one with lowest begin_number)
+
+		let begin_number = backend
+			.block_number_from_id(&BlockId::Hash(begin))?
+			.ok_or_else(|| HandleRequestError::InvalidRequest("Missing start block".to_string()))?;
+
+		if begin_number > backend.info().finalized_number {
+			return Err(HandleRequestError::InvalidRequest(
+				"Start block is not finalized".to_string(),
+			));
+		}
+
+		let canon_hash = backend.hash(begin_number)?.expect(
+			"begin number is lower than finalized number; \
+			 all blocks below finalized number must have been imported; \
+			 qed.",
+		);
+
+		if canon_hash != begin {
+			return Err(HandleRequestError::InvalidRequest(
+				"Start block is not in the finalized chain".to_string(),
+			));
+		}
+
+		let mut proofs = Vec::new();
+
+		for (_, last_block) in set_changes.iter() {
+			if proofs.len() >= MAX_CHANGES_PER_WARP_SYNC_PROOF {
+				break;
+			}
+
+			if *last_block <= begin_number {
```

Comment:
```suggestion
			if *last_block < begin_number {
```

I just tested this on smoldot, and we need this line to be changed. Not entirely sure what's happening here, but generally it's better to send the same epoch at the end of one warp sync response, and the start of another, than to risk skipping over one and breaking verification.

Reply:
This code assumes that if you ask for a warp sync proof starting at block `begin` then you know that that block is finalized. If we change this then most likely you'll end up verifying the same proof twice.

Not sure if it's related to this: https://github.com/paritytech/smoldot/blob/c7c6d56521524721463877aae3280c2f44ba2ee5/src/chain/sync/grandpa_warp_sync.rs#L432-L438

But this assumption is no longer valid now as when the warp sync is done (i.e. we can't prove anything further) we should provide an empty response. This is probably an artifact of how the previous code generated the proofs.

Reply:
Why is it beneficial to send the same epoch?
If the requester says "I want a warp sync proof starting from block A", I don't really see the advantage for the answerer to send back A.

Reply:
> Why is it beneficial to send the same epoch?

Other than 'because otherwise verification fails with a bad signature', I'm not sure.